ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/commons/objectivefunctionevaluator/MnecViolationCostEvaluator.java

```diff
@@ -35,62 +41,27 @@ public MnecViolationCostEvaluator(Set<FlowCnec> flowCnecs, Unit unit, FlowResult
     }
 
-    private double computeCost(FlowResult flowResult, FlowCnec mnec) {
-        double initialMargin = initialFlowResult.getMargin(mnec, unit);
-        double currentMargin = flowResult.getMargin(mnec, unit);
-        return Math.max(0, Math.min(0, initialMargin - mnecAcceptableMarginDecrease) - currentMargin);
-    }
-
-    @Override
-    public Pair<Double, List<FlowCnec>> computeCostAndLimitingElements(FlowResult flowResult, Set<String> contingenciesToExclude) {
-        if (Math.abs(mnecViolationCost) < 1e-10) {
-            return Pair.of(0., new ArrayList<>());
-        }
-        double totalMnecMarginViolation = 0;
-        List<FlowCnec> costlyElements = getCostlyElements(flowResult, contingenciesToExclude);
-        for (FlowCnec mnec : costlyElements) {
-            Optional<Contingency> contingency = mnec.getState().getContingency();
-            if (mnec.isMonitored() && (mnec.getState().getContingency().isEmpty() || contingency.isPresent() && !contingenciesToExclude.contains(contingency.get().getId()))) {
-                totalMnecMarginViolation += computeCost(flowResult, mnec);
-            }
-        }
-        return Pair.of(mnecViolationCost * totalMnecMarginViolation, costlyElements);
-    }
-
     @Override
-    public Unit getUnit() {
-        return unit;
-    }
-
-    private List<FlowCnec> getCostlyElements(FlowResult flowResult, Set<String> contingenciesToExclude) {
-        List<FlowCnec> sortedElements = flowCnecs.stream()
-            .filter(cnec -> cnec.getState().getContingency().isEmpty() || !contingenciesToExclude.contains(cnec.getState().getContingency().get().getId()))
-            .filter(Cnec::isMonitored)
-            .collect(Collectors.toMap(
-                Function.identity(),
-                cnec -> computeCost(flowResult, cnec)
-            ))
-            .entrySet().stream()
-            .filter(entry -> entry.getValue() != 0)
-            .sorted(Comparator.comparingDouble(Map.Entry::getValue))
-            .map(Map.Entry::getKey)
-            .collect(Collectors.toList());
-        Collections.reverse(sortedElements);
-
-        return sortedElements;
+    public CostEvaluatorResult evaluate(FlowResult flowResult, RemedialActionActivationResult remedialActionActivationResult) {
+        Map<FlowCnec, Double> costPerMnec = flowCnecs.stream().filter(Cnec::isMonitored).collect(Collectors.toMap(Function.identity(), mnec -> computeMnecCost(flowResult, mnec)));
+        Map<State, Set<FlowCnec>> mnecsPerState = groupFlowCnecsPerState(costPerMnec.keySet());
+        Map<State, Double> costPerState = mnecsPerState.keySet().stream().collect(Collectors.toMap(Function.identity(), state -> Math.abs(mnecViolationCost) < 1e-10 ? 0.0 : mnecViolationCost * mnecsPerState.get(state).stream().mapToDouble(costPerMnec::get).sum()));
+        List<FlowCnec> sortedMnecs = sortFlowCnecsByDecreasingCost(costPerMnec);
+        return new SumCostEvaluatorResult(costPerState, sortedMnecs);
     }
```

Comment:
use the same structure as for loopflows (just the excess method will look different because of margins vs thresholds)

Reply:
I don't think it's worth making an abstract class for now. But use the same structure with the log. And also the same way of computing the costPerState (I'm not sure why the mnec one requires checking the value of mnecViolationCost but not the loopflow one)